### DISCUSSION TITLE
Is this way of using Server Actions encouraged?

### DISCUSSION THREAD

**Discussion by brain1401** [vercel/next.js#58397] [Help] [answered]
### Summary

I did not use server actions in the Form, but instead I used server actions with useEffect and then set the values with setState. and everything seems to work fine. the value of problem state was correctly logged to console.log.

```tsx
"use client";

import { myAction } from '@/serverActions/test';
import { Problem } from '@/types/problems';
import {useState, useEffect} from 'react';


export default function TestPage() {
  const [problem, setProblem] = useState<Problem>();

  useEffect(() => {

    const fetchData = async () => {
      const result = await myAction("201d50a2-f376-45de-87a0-c76d535d44b8");
      setProblem(result);
    }

    fetchData();
  },[])

  useEffect(() => {
    console.log(problem);
  },[problem])

  return (
    <div>page</div>
  )
}
```

**Comment by wlechowicz** [comment, marked as answer]
Well this is just the traditional "fetching in an effect" with all its problems. What if, by the time `myAction` resolves, the `TestPage` component has already been unmounted? `setProblem` in the closure becomes a memory leak and will throw things at you in the console, as you're trying to update a state that doesn't exist anymore. Since you don't directly control the `fetch` that's happening behind the scenes of your server action's server-to-client boundary crossing, you can't cancel it. You'd have to add a `gotCanceled` flag in the cleanup function of the effect and skip the state update if it's true. But you're still wasting the fetch.

Long story short, IMO zero wins by using an action like this, might as well (or actually, for the better) fetch an API route and be explicit about it. At least it will make you and everyone else reading the code aware that input to the action must be validated like any incoming request body.
